app/src/org/commcare/dalvik/geo/HereFunctionHandler.java

```diff
@@ -0,0 +1,149 @@
+package org.commcare.dalvik.geo;
+
+import android.Manifest;
+import android.content.Context;
+import android.content.pm.PackageManager;
+import android.location.Location;
+import android.location.LocationListener;
+import android.location.LocationManager;
+import android.os.Bundle;
+import android.os.Looper;
+import android.support.v4.content.ContextCompat;
+import android.util.Log;
+
+import org.commcare.dalvik.activities.EntitySelectActivity;
+import org.commcare.dalvik.application.CommCareApplication;
+import org.javarosa.core.model.condition.EvaluationContext;
+import org.javarosa.core.model.condition.IFunctionHandler;
+import org.javarosa.core.model.data.GeoPointData;
+import org.odk.collect.android.utilities.GeoUtils;
+
+import java.util.Set;
+import java.util.Vector;
+
+/**
+ * HereFunctionHandler is a class that can be used to evaluate the here() XPath expression.
+ * In any context where here() needs to be evaluated, create a HereFunctionHandler and
+ * call addFunctionHandler to add it to the XPath evaluator.
```

Comment:
My approach when writing comments is to:
 - be concise. "HereFunctionHandler is a class that can be" can be inferred by the surrounding context
 - avoid describing how code is used. "call addFunctionHandler to add it ..." might get stale if we change the code but forget to update this comment

My preference would be something along the lines of "Allows evaluation contexts to support the here() XPath function, which returns the current location."

Feel free to ignore this and write comments however you feel proper; just wanted let you what I like :)